Foundation/JSONSerialization.swift

```diff
@@ -28,6 +28,253 @@ extension JSONSerialization {
+extension Double : _JSONValue {
+    
+    @_specialize(where Visitor == JSONWriter)
+    @_specialize(where Visitor == JSONValidator)
+    fileprivate func visit<Visitor: JSONVisitor>(with visitor: inout Visitor) throws {
+        guard isFinite && !isNaN else {
+            throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.propertyListReadCorrupt.rawValue, userInfo: ["NSDebugDescription" : "Invalid number value (\(self)) in JSON write"])
+        }
+        try visitor.visit(self)
+    }
+}
+
+extension Bool : _JSONValue {
+    
+    @_specialize(where Visitor == JSONWriter)
+    @_specialize(where Visitor == JSONValidator)
+    fileprivate func visit<Visitor: JSONVisitor>(with visitor: inout Visitor) throws {
+        try visitor.visit(self)
+    }
+}
+
+extension Decimal : _JSONValue {
+    
+    @_specialize(where Visitor == JSONWriter)
+    @_specialize(where Visitor == JSONValidator)
+    fileprivate func visit<Visitor: JSONVisitor>(with visitor: inout Visitor) throws {
+        guard isFinite else {
+            throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.propertyListReadCorrupt.rawValue, userInfo: ["NSDebugDescription" : "Invalid number value (\(self)) in JSON write"])
+        }
+        try visitor.visit(self)
+    }
+}
+
+extension NSNull : _JSONValue {
+    
+    @_specialize(where Visitor == JSONWriter)
+    @_specialize(where Visitor == JSONValidator)
+    fileprivate func visit<Visitor: JSONVisitor>(with visitor: inout Visitor) throws {
+        try visitor.visit(self)
+    }
+}
+
+extension NSNumber : _JSONValue {
+    
+    @_specialize(where Visitor == JSONWriter)
+    @_specialize(where Visitor == JSONValidator)
+    fileprivate func visit<Visitor: JSONVisitor>(with visitor: inout Visitor) throws {
+        if CFNumberIsFloatType(_cfObject) {
+            try doubleValue.visit(with: &visitor)
+        } else {
+            switch _cfTypeID {
+            case CFBooleanGetTypeID():
+                try boolValue.visit(with: &visitor)
+            default:
+                try intValue.visit(with: &visitor)
+            }
+        }
+    }
+}
+
+extension NSArray : _JSONValue {
+    
+    @_specialize(where Visitor == JSONWriter)
+    @_specialize(where Visitor == JSONValidator)
+    fileprivate func visit<Visitor: JSONVisitor>(with visitor: inout Visitor) throws {
+        try visitor.visit(self) //typecheck lazily as we go
+    }
+}
+
+extension NSDictionary : _JSONValue {
+    
+    @_specialize(where Visitor == JSONWriter)
+    @_specialize(where Visitor == JSONValidator)
+    fileprivate func visit<Visitor: JSONVisitor>(with visitor: inout Visitor) throws {
+        try visitor.visit(self) //typecheck lazily as we go
+    }
+}
+
+extension NSString : _JSONValue {
+    
+    @_specialize(where Visitor == JSONWriter)
+    @_specialize(where Visitor == JSONValidator)
+    fileprivate func visit<Visitor: JSONVisitor>(with visitor: inout Visitor) throws {
+        try visitor.visit(self as String)
```

Comment:
Bridging here could be done lazily as well at the cost of having yet another override point on JSONVisitor. Dunno if it's worth it.